### DISCUSSION TITLE
Sporadic Occasional Buzz Glitching

### DISCUSSION THREAD

**Discussion by Neurosmith24** [kriswilk/buzzah#20] [General]
We're having a random, occasional issue with the glove buzzing several times now. At some point into a session (ranging from 2 minutes in to an hour into a session), the random alternating buzzing of the LRAs suddenly stops, replaced by a constant buzz on only one finger (initially the same finger each time). When this happens all of the drivers still show green lights. Most of the time it has happened I've powered off and then right back on and the issue has resolved. 

My initial theory was that maybe the electronics were overheating. They were all in a plastic box with some foam padding, and that was inside of a little pack. I since took them out of the box and just have the electronics attached a piece of cardboard. But the same issue happened a few days ago

My next theory was that maybe the AMPLITUDE_MAX was too high and somehow overdoing something and throwing some kind of error. So, I reset AMPLITUDE_MAX to 100 (was at 127).

I also swapped the drivers and corresponding wiring for two drivers - the one that would consistently remain buzzing when it glitched swapped with the finger next to it - right middle.

Tested that for 3 sessions yesterday and no glitches. Then today, 2 minutes in, it did it again. This time a different finger, right pinky was the one that remained buzzing. Rebooted and it worked - for another minute then did it again. Rebooted again and so far so good 10 minutes into the session

Any other ideas why this might be happening?  Seems unlikely that it would be a software issue, but @kriswilk, any insight into what, if anything, in the code might throw that kind of error? So, wondering if it might be one of the drivers glitching? The mux? The MC?  @PWPInnovator898 have you come across any issues like this?

The parameters I'm using are:
ACTUATOR_TYPE = "LRA"
# Maximum drive voltage, in V
ACTUATOR_VOLTAGE = 2.5 * math.sqrt(2)
# Driving frequency, in Hz
ACTUATOR_FREQUENCY = 175
...
AMPLITUDE_MIN = 100
AMPLITUDE_MAX = 127

I've reset # Vibration amplitude range to:
AMPLITUDE_MIN = 100
AMPLITUDE_MAX = 100

Any thoughts or insights appreciated.

**Comment by Neurosmith24** [comment]
I should also mention that I wired the "5V" pad from the MC board to the "V+" pad on the multiplexer and switched the multiplexer's the switch to the "V+" to drive all of its output ports with a 5V power level. It's been running fine like this for several weeks.

**Comment by kriswilk** [comment]
A problem that happens so randomly/intermittently suggests a hardware issue. The first two possibilities that I thought of:

- Power supply brownout: If the input voltage sags due to discharge or high current draw, the processor could malfunction and get locked up. Make sure the USB power supply is good and the connection is firm.
- Poor connection to a driver: Upon startup, the main board tests for the presence of each driver. Those that are missing are ignored. But those found will be expected to be there during later execution. If a driver becomes disconnected, it will cause the code to fail when it tries to address it.

When the lockup happens, do you get a red light on the main board? You may want to try connecting to the "serial console" on the QT Py board. It is a command line interface to the QT Py board that you can access by connecting to your PC. It may show one or more errors when the problem happens and could help to diagnose the root cause.

To see the serial console output, refer to the following page...it suggests a program called "Mu", which is a simple code editor and serial console all in one package.

https://learn.adafruit.com/welcome-to-circuitpython/kattni-connecting-to-the-serial-console

**Comment by Neurosmith24** [comment]
Thank you @kriswilk ! Good insights on the power supply that I hadn't thought of and I'll look into the serial console. I haven't yet seen a red light on the main board when it happens.

**Comment by Neurosmith24** [comment]
Hi Kris, is the main board supposed to have a green light always on when the code is running? Mine blinks a few times on startup then isn't lit. Attached video shows starting up.

https://github.com/kriswilk/buzzah/assets/5769522/bddaf9a4-14da-4eb7-a316-cc1279542c90

**Reply by kriswilk** [reply]
That's the normal behaviour. The lights on the mux and drivers should be lit but the QT Py board will only blink on startup. On any error it will be solid red.

**Comment by Neurosmith24** [comment]
Hi Kris,

So last night we ran the gloves again and they glitched and the red light on the main board did light up (we ran it earlier in the day for 90 minutes with no issue at all). We tried rebooting 4 times and each time it glitched, with a different finger remaining buzzing each time, but always on the right hand.

So this morning I connected to the serial console on the board, I opened up Mu and connected USB from the main board to computer. This time around I got a strong strange buzz right away on all fingers before I could even click into Mu and click the Serial button to split the screen. So nothing shows up. But then within a second or two the board and gloves resolve and revert to normal buzzing. Tried this several times this morning with the same result. 

I also tried disconnecting a Stemma cable between the Mux and a driver to see if I could force the same error, but when I did that the buzzing just stopped - different result from the behaviour I've been experiencing. 

Would it make sense to try disconnecting the wire from the "5V" pad from theon the MC board to the "V+" pad on the multiplexer and switching the mux  switch back?

Will try just running it connected to the computer in case it throws an error again.

Any of this provide any new clues or ideas? 

Thanks!

**Reply by Neurosmith24** [reply]
Just tried it again and managed to get the following display in the console:
Code stopped by auto-reload. Reloading soon.
soft reboot

It did the same thing. Within a second of starting, it does an odd long buzz on all of the fingers at once for a second or two then resolves.

When I plug it into a battery it doesn't do the odd glitch at the outset that I'm seeing this morning - only when it's plugged into the computer.

**Reply by TactileDesign** [reply]
Hi Neurosmith...  I am just getting started with this system, but one thing that I have learned from Kris is that it takes a lot of force to seat those tiny connectors.  When I first joined all my sockets, I thought I had them all seated firmly.  But after Kris's suggestion, I found that I still had a lot of travel left in all of them!  I would suggest using needle-nose pliers to squeeze the two sockets together after you have put them together with your fingers to see if you can get your connections any tighter.

**Reply by Neurosmith24** [reply]
Thanks, good suggestion. I've visually checked them all and pushed with my fingers as well. One possible next step for me is to just disassemble the whole thing and put it back together. But, today it seems to be running fine. I've done a few stress tests with a new battery and so far so good. And when I run powered by my computer, after an initial weird buzzing it seems to work fine. All of this is making me wonder if it's a power issue.

**Reply by TactileDesign** [reply]
Sounds like things are going well...  A few more points:
1.  I should have mentioned that - if you ever do use needle nose to tighten, make sure power is off since the tips could short out connections.  : )  You probably already knew that - but I wanted to include that comment for future generations who might someday read our erudite pontifications...
2.  When I am plugged into the computer and make a change to the code, my buzzers exhibit a long buzz together before starting their routine.  
3.  I looked at the acoustical response of the three LRAs you have also tested:  The original VG0832022D (small cylinder), the VLV101040A (The Square), and the VG1040003D-ND (the larger cylinder).   I think it is best to drive at 235-250 HZ because there is less bone penetration/conduction and less wave travel along the skin surface from the source point on the fingertip at higher frequencies.  Even up at 235HZ (away from resonance for the two new LRAs), The Square and Large Cylinder both significantly beat the small cylinder in terms of response amplitude and rise/fall times.  The Square wins slightly over the Large Cylinder in amplitude.  I will post scope traces when I get a chance. I am moving forward with my build using the Large Cylinder because I want to minimize contact area with my finger (since I am not using an enclosure for the buzzer like you do).

**Reply by TactileDesign** [reply]
Question - have you been using the 32 gauge wire to solder to the pads as suggested in the manufacturer specs of the LRAs?  I got some, but I feel like I am soldering hairs!  I worry those flimsy wires wont stand up to the stress of me putting my gloves on and off....

**Reply by Neurosmith24** [reply]
Aha! What you describe in item 2 is exactly what I'm getting when I power from the computer. Maybe it's a feature not a bug! 

I use the VLV101040A LRAs (that's the only one I've tried) and soldered 28 gauge wire to them - I missed the mfg specs on that I guess. Wonder if it makes any significant difference? 28 gauge was tough enough to solder - I imagine 32 must be brutal. If it were to make a significant difference though I would get 32 gauge and do them over.

**Reply by TactileDesign** [reply]
The 32 gauge recommendation rationale is apparently because soldering heavier gauge may cause those little pads to lift up off their platform - I a guessing from the added mass.  Given that you got by with heavier and have been using for a while, gives me confidence to move to heavier wire to prevent wire breakage.  I will purchase some extra LRAs in case of "pad lift" down the road.  Either way, they dont last forever...  I thought of something else for you...  As I tinkered around with my samples, I noticed that with a z-axis buzzer, it is easy to choke the vibration if you mechanically smother the buzzer with a holder that does not allow it to vibrate in the z direction - the large cylinder LRA was more prone to this issue than your square - but I think it is true to some extent with the square too.  I am not sure what kind of backing you have to hold the buzzer in place - but I think the best backing would be some kind of spring, or springy cushion - that way you could get maximum vibration out of the LRA at any given setting.

**Reply by Neurosmith24** [reply]
I have the LRA mounted on, and surrounded by foam to try to dampen non-Z axis vibrations. Haven't had any issue with the pads lifting off that I've noticed but I'll inspect. I have more LRAs arriving today so I'll be swapping some out in the next few days - some LRAs vibrate more strongly than others and I want only the strongest ones. 

Here's a pic of my LRAs in their housings.
<img width="698" alt="Screenshot 2024-06-21 at 9 51 29 AM" src="https://github.com/kriswilk/buzzah/assets/5769522/138f3f8d-853f-4e01-b66c-2274d0c9cd57">

**Reply by kriswilk** [reply]
Just had a chance to sit and read your progress. The strange buzzing when you connect to a PC probably has to do with the uC board being enumerated on the USB bus and CircuitPython reloading the code after the board is recognized as a disk drive by the computer's OS.

If you do encounter the lockup again and if the red light turns on (usually it would blink every few seconds), then there should be one or more errors emitted to the serial console. That would be the only hint about the nature of the problem.

I tested disconnecting the mux while in operation and also observed that the uC board did NOT error out (no red light). It's possible that it just ends up in an infinite loop trying to connect rather than throwing an error.

Another consideration in all of this is the peak current consumption. When connected to a PC, be aware that most USB ports can only supply 500 mA. Depending on your actuator and drive settings, you might be close to or beyond that. So the behaviour of the system may be different that when powered by a USB power pack simply for this reason.

**Reply by Neurosmith24** [reply]
A new clue tonight maybe. Been running the gloves powered by the computer the last few days with no issues. However, tonight had several that were a bit different than others - gloves just stopped buzzing completely. I had Mu open and it output the following:

Traceback (most recent call last):
  File "code.py", line 107, in <module>
  File "code.py", line 69, in buzz_sequence
  File "code.py", line 63, in fingers_off
  File "adafruit_drv2605.py", line 247, in realtime_value
  File "adafruit_drv2605.py", line 131, in _write_u8
  File "adafruit_tca9548a.py", line 80, in writeto
OSError: [Errno 19] No such device

Code done running.

Any new insight from this?

Thanks!

**Reply by TactileDesign** [reply]
Hi Neurosmith24,

One thing that occurs to me is the possibility that simply having your RP processor connected to a computer during a run might itself introduce issues that have nothing to do with your Buzzah system.  You are giving your RP a parttime job as as a thumb drive when you attach it to your computer while it is also doing its main job to run your Buzzah.  So the two jobs might have conflicts occasionally.  It also seems to me that having the code and libraries actively accessible to your computer during a therapy session opens the RP software up to getting "messed with" by your computer - especially with Mu open, a program designed to reprogram and interact with the RP.  In fact, I was surprised that I could even open the CODE with Mu while the process was running without getting into trouble.  I have noticed the light on my thumb drives light up at odd times when idle in a computer when nothing should be going on - I have always assumed it was my computer doing some kind of house cleaning - so one of those kinds of things could be getting you.  And there is always the possibility your computer could have some kind of virus trying to write to your thumbdrive.  I am sure if someone was running with a DBS system attached to their brain, the manufacturers would not want the system to be attached to a personal computer that could muck things up...   Secondly - If everything works fine if you power using a USB battery pack, why dont you just use a battery pack instead of your computer?  Are you doing some kind of Buzzah data logging with your computer?

**Reply by Neurosmith24** [reply]
Thanks. Most of the problems happened with a battery pack. In fact, last night was the first time they happened while hooked up to the computer. They initially started while using battery packs. When I switched to the computer I wasn't seeing any glitches for several sessions, so I assumed the battery packs were part of the issue.

**Reply by kriswilk** [reply]
@TactileDesign I had suggested that @Neurosmith24 run the system from his PC for a while to see if, when one of his sporadic failures occurred, any helpful errors were emitted to the serial console.

@Neurosmith24 The traceback you copied does help. It indicates that the problem is with the core communication between the QT Py board and the mux board. Specifically, when an instruction is being sent to the mux board for delivery to one of the drivers, an error is being returned that the specified device doesn't exist. Without looking more deeply I can't tell if the "device" is the mux board itself that isn't reachable, or a specific driver (but I suspect the latter).

The good news is that this is outside the code that I've written. But I'm afraid that's also the bad news... it's not likely a software bug at all.

Although it's possible for a hardware component to be malfunctioning, most of these boards are quite resilient and that would be the last thing I'd consider. It's probably something else.

For instance... how long are the cables connecting QT Py to mux and mux to drivers? The signaling along these wires (the "I2C bus") is designed to go between chips on a single circuit board. Wire lengths measured in inches deteriorate the quality of the signals and might cause intermittent communication problems.

I will try to think of other possibilities but that's one that came to mind.

**Reply by kriswilk** [reply]
I should add that @TactileDesign, your thought about the USB connection adding load to the system is logical. However, that communication is part of the low-level code running the QT Py board and actually doesn't consume many resources at all. It's unlikely to have any effect.

When connected to a PC it IS possible to cause the hardware to reset if you do anything that modifies the contents of the USB drive. But unless you actively edit/save a file on the drive, it should just run the code as usual, forever.

**Comment by Neurosmith24** [comment]
@kriswilk the reference to adafruit_tca9548a.py in the error log I get I think refers to code that the multiplexer uses. Is that correct? Then the following line says "OSError: [Errno 19] No such device." Is that saying that it can't find the multiplexer?

I think you recently updated the lib folder. Should I load the latest version? Maybe that might help?

**Reply by kriswilk** [reply]
I just posted a reply at the same time as you. Yes, the traceback drills down to the mux code but without looking more carefully I don't know if the error means the mux couldn't be reached or if an instruction sent to it couldn't be delivered to a connected device. Either way, there is a connection issue. If it's not a physical one (bad connector, loose wire, etc.), then my only other thought was that the wires you're using might be long enough that the I2C bus is glitching. Having longer wires running from drivers to actuators is OK, but between QT Py -> mux -> drivers should be kept as short as possible... no more than a few inches each.

**Reply by kriswilk** [reply]
BTW, yes, I updated the lib folder but if you didn't have the right files for your version of CircuitPython, the whole thing wouldn't run.

That said, you can certainly upgrade your QT Py to the latest CircuitPython (9.0.5) and then copy the latest lib files. Perhaps it will help, but I kind of doubt it.

**Reply by Neurosmith24** [reply]
I'm just using the Stemma cables, 100mm long, to connect the QT Py to mux and mux to drivers. 

If I decide to update to latest version of CircuitPython, can I just copy the new UF2 file over or do I need to: HOLD the button CLOSEST to the USB port.
WHILE HOLDING the first button, press and release the other button.
CONTINUE TO HOLD the first button until your computer recognizes a new USB drive called "RPI-RP2".
RELEASE the first button.

and then copy it over?

Also, any chance the added wire from the 5V pad on QT Py to the mux V+ might be part of the issue?

**Reply by kriswilk** [reply]
To update CP you need to do the button-pressing routine to get the RPI-RP2 drive to appear. That's how you update the low-level firmware. After that, the board will reboot and you can access the "usual" drive and copy the new "lib" folder over.
